src/i18n.ts

```diff
@@ -0,0 +1,280 @@
+import {
+  Context,
+  Fluent,
+  type FluentBundleOptions,
+  FluentType,
+  type FluentValue,
+  type LocaleId,
+  type MiddlewareFn,
+  type NextFunction,
+  resolve,
+  type TranslationContext,
+} from "./deps.ts";
+
+import { readLocalesDir } from "./utils.ts";
+
+import type {
+  I18nConfig,
+  I18nContextFlavor,
+  TranslateFunction,
+} from "./types.ts";
+
+class FluentContext extends FluentType<string> {
+  constructor(value: string, private defaultValue = "") {
+    super(value);
+  }
+
+  // deno-lint-ignore no-explicit-any
+  toString(scope: any): string {
+    let ctx = JSON.parse(scope.args?.ctx as string);
+    const keys = this.value.split(".");
+    for (const key of keys) {
+      ctx = ctx[key];
+      if (ctx === undefined) break;
+    }
+    return ctx ?? this.defaultValue;
+  }
+}
+
+/**
+ * A custom Fluent function for accessing `Context` properties directly from
+ * Fluent translation files.
+ */
+export function CTX(args: FluentValue[]) {
+  if (args[1] && typeof args[1] !== "string") {
+    throw new TypeError("CTX: Only string type is allowed for defaultValue.");
+  }
+
+  const key = args[0];
+  const defaultValue = typeof args[1] === "string" ? args[1] : undefined;
+
+  if (typeof key === "string") {
+    return new FluentContext(key, defaultValue);
+  }
+
+  throw new TypeError(
+    "CTX: Only type string is allowed for key.",
+  );
+}
+
+export class I18n<C extends Context = Context> {
+  private config: I18nConfig<C>;
+  readonly fluent: Fluent;
+  readonly locales = new Array<string>();
+
+  constructor(config: Partial<I18nConfig>) {
+    this.config = {
+      defaultLocale: "en",
+      fluentBundleOptions: { functions: {} },
+      ...config,
+    };
+
+    if (
+      this.config.fluentBundleOptions &&
+      this.config.fluentBundleOptions.functions
+    ) {
+      this.config.fluentBundleOptions.functions["CTX"] = CTX;
+    } else {
+      this.config.fluentBundleOptions = {
+        ...this.config.fluentBundleOptions,
+        functions: { CTX },
+      };
+    }
+
+    this.fluent = new Fluent(this.config.fluentOptions);
+
+    if (config.directory) {
+      this.loadLocalesDir(config.directory);
```

Comment:
The `deno.ts` I'm talking about is in `examples/deno.ts`.

If I run `deno.ts` from the root repo, start `deno run --allow-all examples/deno.ts` at root repo `/`, it will fire this error:

```
error: Uncaught NotFound: No such file or directory (os error 2), readdir 'locales'
  for (const entry of Deno.readDirSync(path)) {
                           ^
    at Object.opSync (deno:core/01_core.js:170:12)
    at Object.readDirSync (deno:deno_runtime-0.68.0/js/30_fs.js:113:17)
    at readLocalesDir (.../dcdunkan-i18n/src/utils.ts:5:28)
    at I18n.loadLocalesDir (.../dcdunkan-i18n/src/i18n.ts:96:24)
    at new I18n (.../dcdunkan-i18n/src/i18n.ts:87:12)
    at .../dcdunkan-i18n/examples/deno.ts:27:14
```

The reason is in line 30 of `deno.ts`:

```
  directory: "locales",
```

This tells that the `deno.ts` should be run in it directory. IMO, this seems to be a not great design. It would be great if we could make `config.directory` relates to the `deno.ts` not to the starting point (where we start the command). What do you guy think about this?

Reply:
I'm not sure about having the `config.directory` related to the file where it is called. I think it might be a lot of work, which is completely not important in my opinion.

But one thing to mention, examples are supposed to be executed from the examples folder. Not the root folder. But yes, it's not always really helpful. That's why we have a `exampleˋ task in deno.jsonc. You can run `deno task example` from root of the repository and it will run the example.

The specified path always been relative to the root afaik. But if this is a good idea and has more upvotes from other participants, we can add this.

Reply:
Not making any statements about the examples in this repo.

It is not a good idea to interpret file paths as relative to a source file. They should be interpreted as relative to the current working directory of the user. This is because the source files might be fetched from remote URLs, thus making relative paths point to locations on the server, rather than to a local directory.